src/DoubleShapeletPsfApprox.cc

```diff
@@ -0,0 +1,467 @@
+            INVALID_MOMENTS
+        );
+    }
+}
+
+namespace {
+
+// An OptimizerObjective that fits the profile of a DoubleShapelet model
+// (see DoubleShapeletPsfApproxAlgorithm::makeObjective for more info.)
+class ProfileObjective : public OptimizerObjective {
+public:
+
+    ProfileObjective(
+        afw::geom::ellipses::Ellipse const & moments,
+        DoubleShapeletPsfApproxControl const & ctrl,
+        afw::image::Image<Scalar> const & image
+    ) : OptimizerObjective(image.getBBox().getArea(), 4),
+        _minRadius(ctrl.minRadius),
+        _maxRadius(ctrl.maxRadiusBoxFraction * std::sqrt(this->dataSize)),
+        _minRadiusDiff(ctrl.minRadiusDiff),
+        _data(ndarray::Array<Scalar,1,1>(ndarray::allocate(this->dataSize))),
+        _arg(ndarray::Array<Scalar,1,1>(ndarray::allocate(this->dataSize)))
+    {
+        // Radius parameters are defined as factors of the moments ellipse, so
+        // we have to scale the constraints the same way.
+        afw::geom::ellipses::Axes axes(moments.getCore());
+        _minRadius /= axes.getB();
+        _maxRadius /= axes.getA();
+        _minRadiusDiff /= axes.getDeterminantRadius();
+        // We now compute most of the exponential argument to the Gaussian function up
+        // front (everything but the factor of 1/r^2), since that won't change as
+        // the parameters change.
+        afw::geom::AffineTransform gt = moments.getGridTransform();
+        GaussianArgFunctor func(_data.shallow(), _arg.shallow(), &gt);
+        applyPixelFunctor(image, func);
+        _normalization = gt.getLinear().computeDeterminant() * shapelet::ShapeletFunction::FLUX_FACTOR /
+            (2.0*afw::geom::PI);
+    }
+
+    virtual void computeResiduals(
+        ndarray::Array<Scalar const,1,1> const & parameters,
+        ndarray::Array<Scalar,1,1> const & residuals
+    ) const {
+        Scalar iAlpha = parameters[0] * _normalization;
+        Scalar oAlpha = parameters[1] * _normalization;
+        Scalar iR2 = parameters[2] * parameters[2];
+        Scalar oR2 = parameters[3] * parameters[3];
+        residuals.asEigen<Eigen::ArrayXpr>() = _data
+            - (iAlpha/iR2)*(_arg/iR2).exp()
+            - (oAlpha/oR2)*(_arg/oR2).exp();
+    }
+
+    virtual bool differentiateResiduals(
+        ndarray::Array<Scalar const,1,1> const & parameters,
+        ndarray::Array<Scalar,2,-2> const & derivatives
+    ) const {
+        ndarray::EigenView<Scalar,2,-2,Eigen::ArrayXpr> d(derivatives);
+        Scalar iR2 = parameters[2] * parameters[2];
+        Scalar oR2 = parameters[3] * parameters[3];
+        d.col(0) = - (_normalization/iR2)*(_arg/iR2).exp();
+        d.col(1) = - (_normalization/oR2)*(_arg/oR2).exp();
+        d.col(2) = -2.0*d.col(0)*(_arg/iR2 + 1.0)*parameters[0]/parameters[2];
+        d.col(3) = -2.0*d.col(1)*(_arg/oR2 + 1.0)*parameters[1]/parameters[3];
+        return true;
+    }
+
+    virtual bool hasPrior() const { return true; }
```

Comment:
Not sure if this separation between having something and saying to have something is a good thing. But assuming it won't be subclassed that much it might not bite us.